src/Symfony/Component/HttpClient/MockHttpClient.php

```diff
@@ -54,18 +64,20 @@ public function __construct($responseFactory = null, string $baseUri = null)
      */
     public function request(string $method, string $url, array $options = []): ResponseInterface
     {
-        [$url, $options] = $this->prepareRequest($method, $url, $options, ['base_uri' => $this->baseUri], true);
+        [$url, $options] = self::prepareRequest($method, $url, $options, ['base_uri' => $this->baseUri], true);
         $url = implode('', $url);
 
-        if (null === $this->responseFactory) {
-            $response = new MockResponse();
-        } elseif (\is_callable($this->responseFactory)) {
-            $response = ($this->responseFactory)($method, $url, $options);
-        } elseif (!$this->responseFactory->valid()) {
+        if ($this->isAllRequestsWasSent()) {
             throw new TransportException('The response factory iterator passed to MockHttpClient is empty.');
+        }
+
+        ++$this->sentRequestsCount;
+        if (\is_callable($this->responseFactory)) {
+            $response = ($this->responseFactory)($method, $url, $options);
         } else {
-            $response = $this->responseFactory->current();
+            $responseFactory = $this->responseFactory->current();
```

Comment:
None of the changes before this line are required, exception the docblock update, isn't it?
If confirmed, please revert.

Reply:
TBH some behavior was changed. Previously if we pass response instance as mock client argument then it was wrapped into array but now into callable. So calling `request` twice for this case previously raise an exception but for the now it is allowed call `request` multiple time and get same request.

This bevavior was changed to be more consistent with callables: when pass single callable - then it will executed for each request, when pass array of callables then it allows call `request` same times how many callables was passed.

Reply:
Does it make sense to return always the same response whatever the number of times the method is called? Don't we want tighter control to be sure we test the behavior we really want to test - and detect any regressions? That's the reasoning with the current behavior. WDYT?

Reply:
You still able to pass `[$response]` to be confident that request performed only once. Return same response make sense for cases when you want mock (not spy) some 3rdparty.

Reply:
I'm not convinced we should change this behavior - especially because it's not the main purpose of the PR. Can you pleas revert this part?

Reply:
Will do. Is there chance that changing behavior could be accepted in separate PR?